### ISSUE TITLE
[member-ordering] Add support to treat member function fields separately from other member fields

### ISSUE THREAD

**Issue by mdziekon** [typescript-eslint/typescript-eslint#1563]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/member-ordering": [
      "error",
      {
        "interfaces": [
          "signature",
          "field",
          "constructor",
          "method"
        ]
      }
    ]
  }
}
```

```TS
interface MyComponentProps {
    propA: string;
    propB: string;
    onActionA: () => void;
    onActionB: () => void;
    propC: string;
}
```

**Expected Result**

(example error message)

> Member propC should be declared before all function field definitions.

**Actual Result**

> No error

**Additional Info**

In the current version of `member-ordering` rule, fields of an interface which are functions, but not member methods, are treated the same as every other type of field (which is quite understandable, these are not member methods after all). However, the pattern presented above is quite commonly used in React's prop types, where we define an interface forcing the component's user to pass some kind of function to the props (eg. an event handler function). For code consistency purposes, it would be nice to be able to enforce separate ordering for such functions, so that, for example, data carrying fields go first, and functions like handlers go next.

Would it be feasible to implement a separate option, eg. `field-functions`, allowing for more granular control for this pattern of usage?

I've seen similar issues related to this rule (eg. #417), however back there the issue is not exactly the same and use cases are quite different. I don't know if there are any other cases besides React where this pattern occurs, but I don't think it's an edge case either.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.18.0` |
| `@typescript-eslint/parser`        | `2.18.0` |
| `TypeScript`                       | `3.7.4` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `10.15.3` |
| `npm`                              | `6.4.1` |

**Comment by bradzacher** [comment]
The rule has full support for class properties declaring a function as their value, and for method signatures in types. Both of these cases are treated as if they were "method"s.

```ts
class Foo {
  methodAsProp = () => 'foo';
}

interface Bar {
  methodSignature(): string;
}
```

However it is not really possible to support the case you want; in that the following should be treated as a method:
```ts
interface Foo {
  propertyWithFunctionType : () => string;
}
```
For simple cases like above, it can work fine.
However once you move to anything even mildly more complex involving type references, this evolves into a problem which requires type information, which the rule does not use.

```ts
// file1.ts
export type ImportedFnType = () => string;

// file2.ts
import { ImportedFnType } from './file1'

type LocalFnType = () => number;

interface Foo {
  one: () => boolean;     // easy to solve for
  two: LocalFnType;       // possible, but hard to solve without type information
  three: ImportedFnType;  // impossible to solve without type information
}
```

We generally avoid adding type information to rules such that they're only used in certain conditions, as a very large portion of our users don't use type information, so adding it to a rule that doesn't 100% need it blocks many users from using the rule.

---

Why do we need to use type information in this case? Why can't we just support the easy, direct method definition case?

Because otherwise this creates a weird dual handling for the rule, and would end up confusing many users.
We don't want to end up with a case where the user says "I know the type of this reference is a function, why is the rule telling me to sort it with the properties!?!?"

We already have seen this to some degree with the handling of class properties with a function value, we've had people complain that `myClassProp = someFunction` should be treated as a method. This has been accepted as a known false-positive, because the vast, **vast**, ***vast*** majority of cases involve inline value declarations, not referenced value declarations.
I don't think the same definitive statement can be said for types, however.

---

As per the first point - the workaround for this is to just use a method signature instead of a typed property.

**Comment by mdziekon** [comment]
Thank you for your explanation, I must admit that I've anticipated the case you've presented and was afraid that it will be a major blocker for this request. Seems like the only way to implement this request (assuming that anyone would actually want to take up this task) would be to create a new, separate rule which would use type information, but obviously a second implementation has downsides on its own, so I don't expect it to be even considered.

As for the workaround, my initial tests on my codebase show that indeed the method signature can be used instead, so I'll see how it works with the rest of the codebase. The downside of this approach is that I have no way to force anyone to only use method signatures, so there can still be some discrepancies in styles.

Anyway, thank you for your time to answer my request.
